### ISSUE TITLE
question about TotalSamplesSeen

### ISSUE THREAD

**Issue by NewEricWang** [Microsoft/CNTK#295]
Hi, 
     I am newbie in using CNTK. I find the value of the parameter "TotalSamplesSeen" is different along each training epoch. How is the value of the parameter "TotalSamplesSeen" computed? Why is it varied? 

Thanks! 

NewEric

**Comment by dongyu888** [comment]
If it’s frame mode they will be the same. If it’s sequence mode, they will be different since the length of the sequences are different.

**Comment by NewEricWang** [comment]
dongyu888,Thanks! 
      Could you tell me how to set the frame mode? I can't find the setup in the example config file and manual book.
      I think I use the frame mode, the config is as follow:
    reader = [
        readerType = "HTKMLFReader"
        readMethod = "blockRandomize"
        miniBatchMode = "partial"
        randomize = "auto"
        verbosity = 1

        features = [
            dim = 660
            type = "real"
            scpFile = "$DataDir$/in.scp"
        ]
        labels = [
            mlfFile = "$DataDir$/in.mlf"
            labelMappingFile = "$DataDir$/state.list"
            labelDim = 134
            labelType = "category"
        ]
    ]

      The training log is as follow: 
Finished Epoch[ 1 of 30]: [Training Set] TrainLossPerSample = 5.5509787; TotalSamplesSeen = 24402857; EvalErrPerSample = 0.78520584;
Finished Epoch[ 2 of 30]: [Training Set] TrainLossPerSample = 5.5491219; TotalSamplesSeen = 48805714; EvalErrPerSample = 0.78492105;
Finished Epoch[ 3 of 30]: [Training Set] TrainLossPerSample = 4.886445; TotalSamplesSeen = 73208571; EvalErrPerSample = 0.71773201;
Finished Epoch[ 4 of 30]: [Training Set] TrainLossPerSample = 2.6311927; TotalSamplesSeen = 97611428; EvalErrPerSample = 0.54962152;
Finished Epoch[ 5 of 30]: [Training Set] TrainLossPerSample = 2.0337906; TotalSamplesSeen = 122014285; EvalErrPerSample = 0.47796956;
Finished Epoch[ 6 of 30]: [Training Set] TrainLossPerSample = 1.7719022; TotalSamplesSeen = 146417142; EvalErrPerSample = 0.43353164;
Finished Epoch[ 7 of 30]: [Training Set] TrainLossPerSample = 1.6770216; TotalSamplesSeen = 170819999; EvalErrPerSample = 0.41748217;
Finished Epoch[ 8 of 30]: [Training Set] TrainLossPerSample = 1.6033801; TotalSamplesSeen = 195222856; EvalErrPerSample = 0.40479535;
Finished Epoch[ 9 of 30]: [Training Set] TrainLossPerSample = 1.542959; TotalSamplesSeen = 219625713; EvalErrPerSample = 0.39410391;
Finished Epoch[10 of 30]: [Training Set] TrainLossPerSample = 1.4919926; TotalSamplesSeen = 244028570; EvalErrPerSample = 0.38488194;

**Comment by dongyu888** [comment]
In the SGD section set

 

frameMode=true

 

However, if you are training RNNs (e.g., LSTM), then frameMode should be set to false since RNNs only make sense to sequences.

 

BTW, the variation does seem to be big in your case even if frameMode is false. However, the training and evaluation criteria seem to be correct to me. So it’s likely that your training data are not well randomized. If this is not the case, there might be bug somewhere and we would like to get your whole repro setup.

 

Thanks,

 

Dong Yu (俞栋)

**Comment by NewEricWang** [comment]
Hi， Prof. Yu,
       Thanks for your instruction. I don't understand the variation that you mentioned above. Is it the variation of EvalErrPerSample or one of TotalSamplesSeen?  Maybe I use a unmatched CV set.  The default value of frameMode is used in my config. 
       Another question is how to use RMB pretraining in CNTK?  I know it is not available in CNTK right now. How do I use a pretraining? Or how do I use the parameters from a trained well model as the initial parameters in CNTK?

      My complete configuration is as follows:
{
RootDir = "."

ConfigDir = "$RootDir$/Config"
DataDir = "$RootDir$/Data"
OutputDir = "$RootDir$/Output"
ModelDir = "$OutputDir$/Models"

# deviceId=-1 for CPU, >=0 for GPU devices, "auto" chooses the best GPU, or CPU if no usable GPU is available
deviceId = auto

command = speechTrain

precision = "float"
traceLevel = "1"
modelPath = "$ModelDir$/cntkSpeechFF.dnn"
parallelTrain = false
stderr = "$OutputDir$/log_train"
#######################################
#  TRAINING CONFIG                    #
#######################################
speechTrain = [
    action = "train"

    SimpleNetworkBuilder = [
        layerSizes = 660:2048*7:134
        trainingCriterion = "CrossEntropyWithSoftmax"
        evalCriterion = "ErrorPrediction"
        layerTypes = "Sigmoid"
        initValueScale=1.0
        applyMeanVarNorm = true
        needPrior = true
    ]

    SGD = [
        epochSize=0
        minibatchSize = 256*5:512*7:1024
        learningRatesPerMB = 1.0*2:0.8*20:0.01
        # autoLearnRateSearchType = true
        numMBsToShowResult = 10
        momentumPerMB = 0.9*10:0.8*10:0.7
        maxEpochs = 30
        dropoutRate=0.0    
        keepCheckPointFiles = false
    
        # Additional optional parameters are: parallelizationStartEpoch
        parallelTrain = [ 
            parallelizationMethod = "DataParallelSGD"
            distributedMBReading = true
    
            # Additional optional parameters are: useZeroThresholdFor1BitQuantization
            dataParallelSGD = [
                gradientBits = 1
            ]
        ]
        autoAdjust=[
            autoAdjustLR="adjustAfterEpoch"
            reduceLearnRateIfImproveLessThan=0.001
            loadBestModel=true
            increaseLearnRateIfImproveMoreThan=1000000000
            learnRateDecreaseFactor=0.6
            learnRateIncreaseFactor=1.382
        ]
    ]
   cvReader=[
       readerType="HTKMLFReader"
       readMethod="rollingWindow"
       miniBatchMode="partial"
       randomize="auto"
       verbosity=1
       features=[
          dim=660
          scpFile="$DataDir$/cv.scp"
       ]
       labels=[
          mlfFile="$DataDir$/cv.mlf"
          labelDim=134
          labelMappingFile="$DataDir$/state.list"
       ]
  ]
}

**Comment by dongyu888** [comment]
I mean the TotalSamplesSeen. It seems there are great variabilities for this value. By default I think frameMode is true and so TotalSamplesSeen should be relatively consistent.

 

CNTK does not support RBM pretraining since it is no longer important esp. if you are using ReLU nodes. You can still use layer-wise discriminative training though (which can be applied to ReLU). There is an example in the speech TIMIT folder.

 

You can load the parameters learned from other models with two approaches: 1. Dump them to a text file and then in the NDL, initialize the parameter from the file. There is an example of doing so in the image ResNet folder. 2. Use model editing language (MEL) to copy values from one model to another. The discriminative training example includes the MEL command. You need to consult CNTK book for other MEL commands.

**Comment by NewEricWang** [comment]
Thanks. I get it!
I find the reason that the value of TotalSamplesSeen is varied in different epoch. The value of TotalSamplesSeen in current epoch is  equal to the value in last epoch plus frame sample number used in current epoch. Is it right? Is TotalSamplesSeen only to log total frame sample number used from the beginning to the current epoch?

Is RBM pretraining also no longer important in speech recognition field? As far as I know, ReLU is rarely used in speech acoustic model.

I think I need to spend more time to read the CNTK book for using MEL.
Thanks again!

**Comment by dongyu888** [comment]
Without batch normalization Sigmoid is as good as ReLU. With BN, ReLU can be trained faster and converge to a better place. This is true for ASR as well.
